### DISCUSSION TITLE
parse n-ary disjunction

### DISCUSSION THREAD

**Discussion by dgel** [tree-sitter/tree-sitter#1607] [Q&A]
I'm trying to parse a language containing constructs like: `a b | c d | e f`. Rather than treat the `|` operator as a binary operator, I'd like to parse this as:

```
(alternative
  (sequence (token) (token))
  (sequence (token) (token))
  (sequence (token) (token))
```

I'd like to do this using a simple grammar, but the following doesn't work, no matter how I configure the precedences:

```
    _expr: $ => choice(                                                 
      $.alternative,                                                    
      $.sequence,                                                       
      $.token,                                                          
    ),                                                                  
                                                                        
    sequence: $ => prec(9, seq(repeat1($._expr), $._expr)),             
    alternative: $ => prec(8, seq(repeat1(seq($._expr, '|')), $._expr)),
    token: $ => /[a-zA-Z0-9_\-]+/,                                      
```

it generates the following output:

```
(source_file [0, 0] - [1, 0]        
  (sequence [0, 0] - [0, 15]        
    (alternative [0, 0] - [0, 13]   
      (sequence [0, 0] - [0, 9]     
        (alternative [0, 0] - [0, 7]
          (sequence [0, 0] - [0, 3] 
            (token [0, 0] - [0, 1]) 
            (token [0, 2] - [0, 3]))
          (token [0, 6] - [0, 7]))  
        (token [0, 8] - [0, 9]))    
      (token [0, 12] - [0, 13]))    
    (token [0, 14] - [0, 15])))     
```

I can resolve the issues by creating different _expr variants for sequence and alternative, but scaling this up creates a rather complicated grammar. Is there a better way of doing this?

**Comment by ahelwer** [comment]
See https://github.com/tree-sitter/tree-sitter/discussions/1456

tl;dr is that I don't think anyone has yet figured out how to do it. I have [something similar](https://github.com/tlaplus-community/tree-sitter-tlaplus/issues/11) in my language and in the end I couldn't come up with an actual use-case that would actually need this as a flat list instead of a nested tree, so I stopped pursuing it.
